bpf/include/bpf/builtins.h

```diff
@@ -6,17 +6,240 @@
 
 #include "compiler.h"
 
+#ifndef __non_bpf_context
+
 #ifndef lock_xadd
 # define lock_xadd(P, V)	((void) __sync_fetch_and_add((P), (V)))
 #endif
 
-#ifndef memset
-# define memset(S, C, N)	__builtin_memset((S), (C), (N))
+#define __it(x, op) (x -= sizeof(__u##op))
+
+static __always_inline __maybe_unused void
+__bpf_memset_builtin(void *d, __u8 c, __u64 len)
+{
+	/* Everything non-zero or non-const (currently unsupported) as c
+	 * gets handled here.
+	 */
+	__builtin_memset(d, c, len);
+}
+
+static __always_inline void __bpf_memzero(void *d, __u64 len)
+{
+#if __clang_major__ >= 10
+	const __u64 s = 0;
+
+	if (!__builtin_constant_p(len))
+		__throw_build_bug();
+
+	d += len;
+
+	switch (len) {
+#define __it_set(a, b, op) (*(__u##op *)__it(a, op)) = b
+	case 96:         __it_set(d, s, 64);
+	case 88: jmp_88: __it_set(d, s, 64);
+	case 80: jmp_80: __it_set(d, s, 64);
+	case 72: jmp_72: __it_set(d, s, 64);
+	case 64: jmp_64: __it_set(d, s, 64);
+	case 56: jmp_56: __it_set(d, s, 64);
+	case 48: jmp_48: __it_set(d, s, 64);
+	case 40: jmp_40: __it_set(d, s, 64);
+	case 32: jmp_32: __it_set(d, s, 64);
+	case 24: jmp_24: __it_set(d, s, 64);
+	case 16: jmp_16: __it_set(d, s, 64);
+	case  8: jmp_8:  __it_set(d, s, 64);
+		break;
+
+	case 94: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_88;
+	case 86: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_80;
+	case 78: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_72;
+	case 70: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_64;
+	case 62: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_56;
+	case 54: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_48;
+	case 46: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_40;
+	case 38: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_32;
+	case 30: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_24;
+	case 22: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_16;
+	case 14: __it_set(d, s, 16); __it_set(d, s, 32); goto jmp_8;
```

Comment:
Would be nice to have a bunch of unit tests with different sizes for this logic, but I'm guessing the `__non_bpf_context` guard will make that complicated...?

Reply:
No, we could. I'll see to add them. The guard is only since the bpf include infra is also included from the map-migrate tool.